services/packagist/packagist-version.service.js

```diff
@@ -90,25 +85,27 @@ class PackagistVersion extends BasePackagistService {
 
   transform({ includePrereleases, json, user, repo }) {
```

Comment:
As noted on `findRelease()`, we should have one method for working out the latest (stable) version. I think most of the logic in this function was there to handle dev branches being mixed in with the releases. Now that they are on a different endpoint `..~dev.json`, I think we can ditch all this can't we (do correct me if I'm wrong)? If so, this simplifies to something like:

```js
const packageSchema = Joi.array().items(
  Joi.object({version: Joi.string().required()})
)

transform({ includePrereleases, json, user, repo }) {
  const versionsData = json.packages[this.getPackageName(user, repo)]
  const versions = versionsData.map( version => version.version)

  if (includePrereleases) {
    return { version: latest(versions) }
  } else {
    const stableVersion = latest(versions.filter(isStable))
    return { version: stableVersion || latest(versions) }
  }
}
```

Then at that point, the logic for identifying the latest release is the same logic as in your `findRelease()` (which we should rename) but with a `includePrereleases` flag.. so we can move this to the base class and call with `includePrereleases=false` when we use it for the license/PHP version badges.

Reply:
Packagist do not have separate tagged version on different branch. Even the main branch is considered a development branch. The `~dev.json` endpoint is useful for accessing information of branch HEAD, regardless if it is tagged or not. The [current version](https://user-images.githubusercontent.com/91274/152666501-b15fcac7-68f5-49ff-a524-a174d4c73249.png) of [documentation](https://packagist.org/apidoc#get-package-metadata-v2) stated:

> The `p2/$vendor/$package.json` file contains only tagged releases. If you want to fetch information about branches (i.e. dev versions) you need to download `p2/$vendor/$package~dev.json`.

Sometimes the tagged pre-release versions (e.g. v0.1.2) of libraries are published and used in production by others. In which case, the latest version would be "not stable" but is not shown on the `~dev.json` endpoint but is still relevant.